python-spec/src/somacore/query/fast_csr.py

```diff
@@ -0,0 +1,300 @@
+                    chunk_list,
+                    data,
+                    indices,
+                    indptr,
+                    row_rng_mask_bits,
+                    job,
+                )
+                for job in range(n_jobs)
+            ]
+        )
+        _finalize_indptr(indptr)
+        return CSRAccumulatorFinalResult(
+            data=data, indptr=indptr, indices=indices, shape=self.shape
+        )
+
+
+@numba.jit(nopython=True, nogil=True)  # type: ignore
+def _accum_row_length(
+    row_length: npt.NDArray[np.int64], row_ind: npt.NDArray[np.int64]
+) -> None:
+    for rind in row_ind:
+        row_length[rind] += 1
+
+
+@numba.jit(nopython=True, nogil=True)  # type: ignore
+def _copy_chunk_range(
+    data_chunk: npt.NDArray[np.number],
+    row_ind_chunk: npt.NDArray[np.signedinteger],
+    col_ind_chunk: npt.NDArray[np.signedinteger],
+    data: npt.NDArray[np.number],
+    indices: npt.NDArray[np.signedinteger],
+    indptr: npt.NDArray[np.signedinteger],
+    row_rng_mask: int,
+    row_rng_val: int,
+):
+    for n in range(len(data_chunk)):
+        row = row_ind_chunk[n]
+        if (row & row_rng_mask) != row_rng_val:
+            continue
+        ptr = indptr[row]
+        indices[ptr] = col_ind_chunk[n]
+        data[ptr] = data_chunk[n]
+        indptr[row] += 1
+
+
+@numba.jit(nopython=True, nogil=True)  # type: ignore
+def _copy_chunklist_range(
+    chunk_list: numba.typed.List,
+    data: npt.NDArray[np.number],
+    indices: npt.NDArray[np.signedinteger],
+    indptr: npt.NDArray[np.signedinteger],
+    row_rng_mask_bits: int,
+    job: int,
+):
+    row_rng_mask = (2**64 - 1) >> row_rng_mask_bits << row_rng_mask_bits
+    row_rng_val = job << row_rng_mask_bits
+    for data_chunk, row_ind_chunk, col_ind_chunk in chunk_list:
+        _copy_chunk_range(
+            data_chunk,
+            row_ind_chunk,
+            col_ind_chunk,
+            data,
+            indices,
+            indptr,
+            row_rng_mask,
+            row_rng_val,
+        )
+
+
+@numba.jit(nopython=True, nogil=True)  # type: ignore
+def _finalize_indptr(indptr: npt.NDArray[np.signedinteger]):
+    prev = 0
+    for r in range(len(indptr)):
+        t = indptr[r]
+        indptr[r] = prev
+        prev = t
+
+
+def _select_dtype(
+    maxval: int,
+) -> Type[np.signedinteger]:
+    """
+    Ascertain the "best" dtype for a zero-based index. Given our
+    goal of minimizing memory use, "best" is currently defined as
+    smallest.
+    """
+    if maxval > np.iinfo(np.int32).max:
+        return np.int64
+    else:
+        return np.int32
+
+
+def _reindex_and_cast(
+    index: pd.Index, ids: npt.NDArray[np.int64], target_dtype: npt.DTypeLike
+) -> npt.NDArray[np.int64]:
+    return cast(
+        npt.NDArray[np.int64], index.get_indexer(ids).astype(target_dtype, copy=False)
+    )
+
+
+def _read_csr(
+    matrix: scd.SparseNDArray, obs_joinids: pa.Array, var_joinids: pa.Array
+) -> Tuple[
```

Comment:
ibid. re using CSRAccumulatorFinalResult. A more general name might make sense for it. `_DataWithIndices`?

Reply:
Same issue - they are completely different matrix formats.  COO vs CSR